### PR TITLE
Diagnostics: Add CPU history

### PR DESCRIPTION
**Pull request by j82w**
# Pull Request Template

## Description

This adds the CPU load history to all diagnostics requests. This information is important to troubleshoot latency and other problems caused by high CPU usage. This is a best effort and will not trace the CPU history if an exception is hit. If some error does happen this should not block users.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

```json
{
    "DiagnosticVersion": "2",
    "Summary": {
        "StartUtc": "2020-06-05T14:56:53.9601828Z",
        "TotalElapsedTimeInMs": 145.179,
        "UserAgent": "cosmos-netstandard-sdk/3.9.1|3.10.0|03|X64|Microsoft Windows 10.0.19041 |.NET Core 4.6.28801.04|",
        "TotalRequestCount": 1,
        "FailedRequestCount": 0
    },
    "Context": [
        {
            "Id": "ItemSerialize",
            "ElapsedTimeInMs": 0.101
        },
        {
            "Id": "ExtractPkValue",
            "ElapsedTimeInMs": 28.4009
        },
        {
            "Id": "AggregatedClientSideRequestStatistics",
            "ContactedReplicas": [
                {
                    "Count": 12,
                    "Uri": "rntbd://127.0.0.1:10253/apps/DocDbApp/services/DocDbServer19/partitions/a4cb495f-38c8-11e6-8106-8cdcd42c33be/replicas/1p/"
                }
            ],
            "RegionsContacted": [
                "https://127.0.0.1:8081/"
            ],
            "FailedReplicas": []
        },
        {
            "Id": "Microsoft.Azure.Cosmos.Handlers.DiagnosticsHandler",
            "HandlerElapsedTimeInMs": 116.54690000000001
        },
        {
            "Id": "SystemInfo",
            "CpuHistory": "(2020-06-05T14:56:40.6078897Z 50.000), (2020-06-05T14:56:50.6084254Z 3.458)"
        },
        {
            "Id": "Microsoft.Azure.Cosmos.Handlers.RetryHandler",
            "HandlerElapsedTimeInMs": 116.54350000000001
        },
        {
            "Id": "Microsoft.Azure.Cosmos.Handlers.RouterHandler",
            "HandlerElapsedTimeInMs": 116.5233
        },
        {
            "Id": "Microsoft.Azure.Cosmos.Handlers.TransportHandler",
            "HandlerElapsedTimeInMs": 1.4924000000000002
        },
        {
            "Id": "Microsoft.Azure.Documents.ServerStoreModel",
            "ElapsedTimeInMs": 116.46820000000001
        },
        {
            "Id": "AddressResolutionStatistics",
            "StartTimeUtc": "2020-06-05T14:56:54.0522007Z",
            "EndTimeUtc": "2020-06-05T14:56:54.0712330Z",
            "ElapsedTimeInMs": 19.0323,
            "TargetEndpoint": "https://127.0.0.1:8081//addresses/?$resolveFor=dbs%2fkdAZAA%3d%3d%2fcolls%2fkdAZAPI0Aag%3d%2fdocs&$filter=protocol eq rntbd&$partitionKeyRangeIds=0"
        },
        {
            "Id": "StoreResponseStatistics",
            "StartTimeUtc": "2020-06-05T14:56:53.9887790Z",
            "ResponseTimeUtc": "2020-06-05T14:56:54.1050021Z",
            "ElapsedTimeInMs": 116.2231,
            "ResourceType": "Document",
            "OperationType": "Create",
            "LocationEndpoint": "https://127.0.0.1:8081/",
            "ActivityId": "2d891c88-7907-4dde-85f0-66cab771df98",
            "StoreResult": "StorePhysicalAddress: rntbd://127.0.0.1:10253/apps/DocDbApp/services/DocDbServer19/partitions/a4cb495f-38c8-11e6-8106-8cdcd42c33be/replicas/1p/, LSN: 14, GlobalCommittedLsn: -1, PartitionKeyRangeId: 0, IsValid: True, StatusCode: 201, SubStatusCode: 0, RequestCharge: 11.05, ItemLSN: -1, SessionToken: -1#14, UsingLocalLSN: False, TransportException: null"
        }
    ]
}
```

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/Handler/DiagnosticsHandler.cs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+            return base.SendAsync(request, cancellationToken);
+        }
+
+        /// <summary>
+        /// This is a helper class that creates a single static instance to avoid each
+        /// client instance from creating a new CPU monitor.
+        /// </summary>
+        private class DiagnosticsHandlerHelper
+        {
+            public static readonly DiagnosticsHandlerHelper Instance = new DiagnosticsHandlerHelper();
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Handler/DiagnosticsHandler.cs:32]: Should this be Lazy? To avoid monitoring when there was never a request?

Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Handler/DiagnosticsHandler.cs:32]: I don't think it's worth optimizing for that scenario. Any user creating a CosmosClient should be doing a request. I can't think of scenario they would create one, but not use it.

Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/Handler/DiagnosticsHandler.cs:35]: My only concern was the monitor running, measuring, when it was not being used.

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/CosmosDiagnosticsTests.cs [hunk 2]
```diff
@@ -283,7 +283,12 @@ public async Task PointOperationDiagnostic(bool disableDiagnostics)
             if (disableDiagnostics)
             {
                 requestOptions.DiagnosticContextFactory = () => EmptyCosmosDiagnosticsContext.Singleton;
-            };
+            }
+            else
+            {
+                // Add 10 seconds to ensure CPU history is recorded
+                await Task.Delay(TimeSpan.FromSeconds(10));
```
[the remark this answers is not in the document]
Comment (j82w) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/CosmosDiagnosticsTests.cs:290]: It's only 1 test and it's a full end to end test. I would rather it wait the 10 seconds to actually test the code then possibly miss something by using reflection to override the value.

**Comment by kirankumarkolli** [comment]
Please run it through with Fabian once.

**Pull request merged**